rfc-128-continuous-deployment.md

```diff
@@ -0,0 +1,223 @@
+
+For the purpose of automating deployments:
+
+- We need to test that an app can respond to a request for each of the machine classes it runs on [[1](https://github.com/alphagov/govuk-app-deployment/blob/d4bb62aa79face46d2e164e316f38f7a64a1b8be/content-store/config/deploy.rb#L3)] [[2](https://github.com/alphagov/govuk-app-deployment/blob/d4bb62aa79face46d2e164e316f38f7a64a1b8be/whitehall/config/deploy.rb#L13)]. This is purely to prove that an instance of the app is running [[1](https://github.com/alphagov/email-alert-api/pull/1346)], and may be covered by other tests.
+
+- We need to test that an app can connect to any systems it uses for data storage and retrieval, such as a database. This usually has production-specific config [[1](https://github.com/alphagov/email-alert-api/blob/1b5f59f3927637afb0c790702f8f0050788edc9d/config/database.yml#L22)] [[2](https://github.com/alphagov/content-store/blob/f4b61d4f2756341b6418085d878ec83b09d70e9c/config/mongoid.yml#L23)] [[3](https://github.com/alphagov/content-publisher/blob/944fe51b4cb8205348e155b965b8bad57adcf4fe/config/storage.yml#L5)], which could be faulty.
+
+- We do not need to test connectivity to other GOV.UK apps, since this should be handled by [our internal API adapters](https://github.com/alphagov/gds-api-adapters/blob/1ceba690edb37295257328d953874224833fcd5e/lib/gds_api.rb) with no configuration inside the repo. An exception to this is [authentication with Signon](https://github.com/alphagov/email-alert-api/blob/1b5f59f3927637afb0c790702f8f0050788edc9d/config/initializers/gds_sso.rb).
+
+With this in mind, an app has "enough" of these tests when:
+
+- It has a test that checks for a successful response to an arbitrary request.
+- It has a test that checks for a successful login with Signon (if applicable).
+- It has a test that checks for connectivity to any read/write dependencies:
+
+  - Databases e.g. MongoDB, Postgres, MySQL, Redis
+  - Queueing systems e.g. RabbitMQ
+  - File systems e.g. AWS S3
+  - Caching systems e.g. Memcached
+
+> Example: [Travel Advice Publisher](https://github.com/alphagov/travel-advice-publisher) has enough of these tests because it has one to check login with Signon [[1](https://github.com/alphagov/smokey/blob/a60df2cde5c886fa6d54e8cd820d90facca89e8e/features/publishing_tools.feature#L106)] and another to check its `/healthcheck` endpoint for connectivity to Redis and MongoDB [[1](https://github.com/alphagov/smokey/blob/a60df2cde5c886fa6d54e8cd820d90facca89e8e/features/travel_advice_publisher.feature)] [[2](https://github.com/alphagov/travel-advice-publisher/pull/979)].
+
+> **Services are special.** We cannot make a request to a service, as it has no web interface. Instead, we will need infer the service is running indirectly, using metrics sent to Graphite in response to [a regular heartbeat](https://github.com/alphagov/publishing-api/blob/a4a054b768b7c98fcd1448f74fe6e975ab69bb2f/config/schedule.rb#L10).
```

Comment:
I agree the proposal isn't ideal, not least because it requires waiting an arbitrary amount of time for new metric datapoints that prove the service is up. In addition, the RabbitMQ check is [no longer required](https://github.com/alphagov/govuk-rfcs/pull/128/commits/e81017e306ec7b8cd1e2eb04280525093b9c0a6f), so we learn relatively little by doing an intrusive check, over the test/production variability of a Rails app. Given this new disadvantage, the suggested approach seems to add little value over a synchronous check. How about this:

> **Services are special.** We cannot make a request to a service, as it has no web interface. Instead, we will update [the deployment script for each service app](https://github.com/alphagov/govuk-app-deployment/blob/efc730f0f101b2e457abc48cbab67982728a8c93/email-alert-service/config/deploy.rb) to check the process is running after the app is restarted.

Reply:
Yes, I'm much happier with that. Thanks.

Reply:
Updated 👍